pkg/codegen/importer/hcl2.go

```diff
@@ -0,0 +1,467 @@
+				simplestElementType = t
+			}
+		}
+		for _, u := range u.(*schema.UnionType).ElementTypes {
+			if simplestElementType == nil || simplerType(u, simplestElementType) {
+				return false
+			}
+		}
+		return true
+	default:
+		return false
+	}
+}
+
+// zeroValue constructs a zero value of the given type.
+func zeroValue(t schema.Type) model.Expression {
+	switch t := t.(type) {
+	case *schema.MapType:
+		return &model.ObjectConsExpression{}
+	case *schema.ArrayType:
+		return &model.TupleConsExpression{}
+	case *schema.UnionType:
+		// If there is a default type, create a value of that type.
+		if t.DefaultType != nil {
+			return zeroValue(t.DefaultType)
+		}
+		// Otherwise, pick the simplest type in the list.
+		var simplestType schema.Type
+		for _, t := range t.ElementTypes {
+			if simplestType == nil || simplerType(t, simplestType) {
+				simplestType = t
+			}
+		}
+		return zeroValue(simplestType)
+	case *schema.ObjectType:
+		var items []model.ObjectConsItem
+		for _, p := range t.Properties {
+			if p.IsRequired {
+				items = append(items, model.ObjectConsItem{
+					Key: &model.LiteralValueExpression{
+						Value: cty.StringVal(p.Name),
+					},
+					Value: zeroValue(p.Type),
+				})
+			}
+		}
+		return &model.ObjectConsExpression{Items: items}
+	case *schema.TokenType:
+		if t.UnderlyingType != nil {
+			return zeroValue(t.UnderlyingType)
+		}
+		return model.VariableReference(Null)
+	}
+	switch t {
+	case schema.BoolType:
+		x, err := generateValue(t, resource.NewBoolProperty(false))
+		contract.IgnoreError(err)
+		return x
+	case schema.IntType, schema.NumberType:
+		x, err := generateValue(t, resource.NewNumberProperty(0))
+		contract.IgnoreError(err)
+		return x
+	case schema.StringType:
+		x, err := generateValue(t, resource.NewStringProperty(""))
+		contract.IgnoreError(err)
+		return x
+	case schema.ArchiveType, schema.AssetType:
+		return model.VariableReference(Null)
+	case schema.JSONType, schema.AnyType:
+		return &model.ObjectConsExpression{}
+	default:
+		contract.Failf("unexpected schema type %v", t)
+		return nil
+	}
+}
+
+// generatePropertyValue generates the value for the given property. If the value is absent and the property is
+// required, a zero value for the property's type is generated. If the value is absent and the property is not
+// required, no value is generated (i.e. this function returns nil).
+func generatePropertyValue(property *schema.Property, value resource.PropertyValue) (model.Expression, error) {
+	if !value.HasValue() {
+		if !property.IsRequired {
+			return nil, nil
+		}
+		return zeroValue(property.Type), nil
+	}
+
+	return generateValue(property.Type, value)
+}
+
+// generateValue generates a value from the given property value. The given type may or may not match the shape of the
+// given value.
+func generateValue(typ schema.Type, value resource.PropertyValue) (model.Expression, error) {
+	switch {
+	case value.IsArchive():
+		return nil, fmt.Errorf("NYI: archives")
```

Comment:
What does `NYI` mean? Not yet implemented?

Reply:
Yes. We shouldn't hit this case in the scope of the `import` command, since none of the providers' `Read` functions will return assets or archives.